_posts/2019-11-13-variational-convnet/variational_convnet.Rmd

```diff
@@ -0,0 +1,619 @@
+  tfd_kl_divergence(q, p) / n
+
+model <- keras_model_sequential()
+model %>% 
+  layer_conv_1d_flipout(
+    filters = 12,
+    kernel_size = 3, 
+    activation = "relu",
+    kernel_divergence_fn = kl_div
+  ) %>%
+  layer_conv_1d_flipout(
+    filters = 24,
+    kernel_size = 5, 
+    activation = "relu",
+    kernel_divergence_fn = kl_div
+  ) %>%
+  layer_conv_1d_flipout(
+    filters = 48,
+    kernel_size = 7, 
+    activation = "relu",
+    kernel_divergence_fn = kl_div
+  ) %>%
+  layer_global_average_pooling_1d() %>% 
+  layer_dense_flipout(
+    units = 48,
+    activation = "relu",
+    kernel_divergence_fn = kl_div
+  ) %>% 
+  layer_dense_flipout(
+    num_classes, 
+    kernel_divergence_fn = kl_div,
+    name = "dense_output"
+  ) %>%
+  layer_one_hot_categorical(event_size = num_classes)
+```
+
+We tell the network to minimize the negative log likelihood.
+
+```{r}
+nll <- function(y, model) - (model %>% tfd_log_prob(y))
+```
+
+This will become part of the loss. The way we set up this example, this is not its most substantial part though. Here, what dominates the loss is the sum of the KL divergences, added (automatically) to `model$losses`.
+
+In a setup like this, it's interesting to monitor both parts of the loss separately. We can do this by means of two metrics:
+
+```{r}
+# the KL part of the loss
+kl_part <-  function(y_true, y_pred) {
+    kl <- tf$reduce_sum(model$losses)
+    kl
+}
+
+# the NLL part
+nll_part <- function(y_true, y_pred) {
+    cat_dist <- tfd_one_hot_categorical(logits = y_pred)
+    nll <- - (cat_dist %>% tfd_log_prob(y_true) %>% tf$reduce_mean())
+    nll
+}
+```
+
+We train somewhat longer than Nick did in the original post, allowing for early stopping though. 
+
+```{r}
+model %>% compile(
+  optimizer = "rmsprop",
+  loss = nll,
+  metrics = c("accuracy", 
+              custom_metric("kl_part", kl_part),
+              custom_metric("nll_part", nll_part)),
+  experimental_run_tf_function = FALSE
+)
+
+train_history <- model %>% fit(
+  train_dataset,
+  epochs = 1000,
+  validation_data = test_dataset,
+  callbacks = list(
+    callback_early_stopping(patience = 10)
+  )
+)
+```
+
+While the overall loss declines linearly (and probably would for many more epochs), this is not the case for classification accuracy and the NLL part of the loss:
+
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = ""}
+knitr::include_graphics("images/history.png")
+```
+
+Final accuracy is not as high as in the non-variational setup, though still not bad for a six-class problem. We see that without any additional regularization, there is very little overfitting to the training data.
+
+Now how do we obtain predictions from this model?
+
+## Probabilistic predictions
+
+Though we won't go into this here, it's good to know that we can't just access the output distributions; through their `kernel_posterior` attribute, we can access the hidden layers' posterior weight distributions as well. 
```

Comment:
```suggestion
Though we won't go into this here, it's good to know that we access more than just the output distributions; through their `kernel_posterior` attribute, we can access the hidden layers' posterior weight distributions as well. 
```

Reply:
^^ Assuming that's what you mean (my original reading of it was you saying _We can't access the output distributions_, until I read the other part of the sentence)

Reply:
oh yeah... you always spot the weird-ish phrases ...